## Ortholog prediction

`/global/projects/programs/source/orthology/orthofinder/2.5.4/orthofinder.py -f /EBB-ZIVNFS/amikhail/termite_genomes/eusociality_paper/Ortho
finder/proteomes`

Inferring HOGs by providing a phylogenetic tree
`/global/projects/programs/source/orthology/orthofinder/2.5.4/orthofinder.py -ft /EBB-ZIVNFS/amikhail/termite_genomes/eusociality_paper/proteomes/OrthoFinder/Results_Dec12_1 -s /EBB-ZIVNFS/amikhail/termite_genomes/eusociality_paper/SpeciesTree_mod.txt`

Extracting sequences of single-copy and all orthologs: `scripts/extractHOGsequences.ipynb`

## Creating and trimming alignments

1. Running PRANK

`sbatch --array=1-826%20 scripts/prank_slurm.sh`

2. Extracting corresponding cds sequences with `scripts/get_cds_sysarg.py`

3. Creating cds alignments with pal2nal

`ls *cds.fasta | sed 's/_cds.fasta//' | while read i; do /global/projects/programs/source/pal2nal.v14/pal2nal.pl prank/${i}-prank.out.best.fas ${i}_cds.fasta -output fasta > ${i}_cds.aln; done`

4. Trimming cds alignments

`ls *_cds.aln | while read i; do /EBB-ZIVNFS/amikhail/termite_genomes/eusociality_paper/src/Gblocks_0.91b/Gblocks $i -t=c -b5=h; done`

`python3 scripts/gblock_editor.py`

5. Creating trees for multi-copy orthologs with `scripts/run_fasttree.sh`

## Running CodeML

Modify ctl file (codeml/null.ctl) by adding the input cds alignment and tree topology (SpeciesTree\_mod.txt)

Parsing output

`grep 'dS tree:' -A 6 OG0006185_free.out | sed -r '/^\s*$/d'`

Then run `scripts/Parse_codeml_trees.ipynb`

## Running RELAX

Tree for testing three origins of TW: `relax/SpeciesTree.txt`. Result table `relax/relax_table.tsv` produced by `scripts/parse_relax_output.ipynb`

Trees for bootstrap analysis: `relax/bootstrap_tree_all` generated by `scripts/relax_bootstrap_trees.py`. Result table `relax/bootstrap_relax.csv` produced by `scripts/parse_relax_output.ipynb` and visualised by `scripts/RELAX_bootstrap.R`. 

## Running aBSREL

Running aBSREL with three origins of TW as foreground branches: `scripts/run_absrel.sh`

Running aBSREL in a test mode: `scripts/run_absrel_testmode.sh`

## Visualisation

dN/dS boxplots per social category together with other test results combined with the phylogeny: `k_RELAX_sociality.R`
